Bluetooth: Host: Issue all auto-initiated commands synchronously

rugeGerritsen opened the pull request:
This solves the following issues:
- Auto-initiation of commands is aborted if one of the issued commands
  fails
- The controller return value for a PHY update or DL update is not
  made available to the application.

It also simplifies the code:
- We no longer need to keep track of if the auto-phy update
  or auto-dl update has completed. If the controller receives
  another LE Set PHY while the procedure is pending in the LL, it can
  decide if wants to accept another procedure initiation or not.
- We no longer need to auto-initiate auto-initiation of commands in
  multiple places

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -956,7 +965,7 @@ static void conn_auto_initiate(struct bt_conn *conn)
 		err = bt_le_set_phy(conn, 0U, BT_HCI_LE_PHY_PREFER_2M,
 				    BT_HCI_LE_PHY_PREFER_2M,
-				    BT_HCI_LE_PHY_CODED_ANY);
+				    BT_HCI_LE_PHY_CODED_ANY, false);
```
I think they should all use send_sync.
If you ignore the HCI error code here, then you wouldn't go to the data length update procedure below. Since that would only continue when the phy update complete event arrives.

rugeGerritsen replied:
@joerchan , I'm not sure if I fully understand the problem. I changed the command to always be synchronous.

joerchan replied:
`hci_le_read_remote_features` is not using the send_sync variant.
If it returns an HCI error code it would lead to the rest of the procedures not being done.

rugeGerritsen replied:
Fixed by issuing all commands synchronously. That brings in other advantages as well.

rugeGerritsen committed:
Bluetooth: Host: Issue all auto-initiated commands synchronously

This solves the following issues:
- Auto-initiation of commands is aborted if one of the issued commands
  fails
- The controller return value for a PHY update or DL update is not
  made available to the application.

As a result, the host no longer prevents the application from calling
bt_conn_le_data_len_update() and bt_conn_le_phy_update() before the
auto-initiated procedure completes. Now the controller may or may not
accept the additional command issued by the application.

It also simplifies the code:
- We no longer need to keep track of if the auto-phy update
  or auto-dl update has completed. If the controller receives
  another LE Set PHY while the procedure is pending in the LL, it can
  decide if wants to accept another procedure initiation or not.
- We no longer need to auto-initiate auto-initiation of commands in
  multiple places

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -923,6 +921,35 @@ static struct bt_conn *find_pending_connect(uint8_t role, bt_addr_le_t *peer_add
+		}
+	}
+#else
+	(void)conn;
```
```suggestion
	ARG_UNUSED(conn);
```

Thalley commented on subsys/bluetooth/host/conn_internal.h:
```diff
@@ -32,20 +32,14 @@ enum {
 	BT_CONN_FORCE_PAIR,             /* Pairing even with existing keys. */
 
-	BT_CONN_AUTO_PHY_COMPLETE,      /* Auto-initiated PHY procedure done */
 	BT_CONN_AUTO_FEATURE_EXCH,	/* Auto-initiated LE Feat done */
 	BT_CONN_AUTO_VERSION_INFO,      /* Auto-initiated LE version done */
```
What about these two?

rugeGerritsen replied:
We need to keep them to prevent the application from obtaining version and feature info in `bt_conn_get_remote_info()` when the feature and version exchange is not completed

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -2408,11 +2403,6 @@ int bt_conn_le_phy_update(struct bt_conn *conn,
 {
 	uint8_t phy_opts, all_phys;
 
-	if (IS_ENABLED(CONFIG_BT_AUTO_PHY_UPDATE) &&
-	    !atomic_test_bit(conn->flags, BT_CONN_AUTO_PHY_COMPLETE)) {
-		return -EAGAIN;
-	}
-
```
I like the idea of this PR, but in the case that the application attempts to do a `bt_conn_le_phy_update` before the auto-initiated PHY update has completed (i.e. before `le_phy_update_complete` has been called), doesn't removing such a check like this cause issues?

joerchan replied:
Yes, this should be kept. If the AUTO_PHY (or the others) are not completed the application should not be able to start these. I think there might be an issue with updating the internal state of the conn object. In addition this would likely result in an HCI error code as well.

rugeGerritsen replied:
I had a thought about that. Before this PR:
```
	if (IS_ENABLED(CONFIG_BT_AUTO_PHY_UPDATE) &&
	    !atomic_test_bit(conn->flags, BT_CONN_AUTO_PHY_COMPLETE)) {
		return -EAGAIN;
	}
```
would result in an error code. Now 
```
	return bt_le_set_phy(conn, all_phys, param->pref_tx_phy,
			     param->pref_rx_phy, phy_opts);
```
would return an error.  

As both code paths will return an error, I thought that it was better to simplify the code. Feel free to disagree.

Thalley replied:
@rugeGerritsen  Can you elaborate on when/why `bt_le_set_phy` will return an error if the stack is doing a PHY update when the application calls `bt_conn_le_phy_update`? I can't find it

rugeGerritsen replied:
`bt_conn_le_phy_update()` issues the command to the Link Layer. Some link layers may simply accept the command, others may return "BUSY". For both cases, I believe the application gets what it wants. Either two phy update callbacks, or only one and an error when calling `bt_conn_le_phy_update()`

Thalley replied:
Ah, so this is an intentional change in host behavior? 

The HCI spec does not disallow this, so that's fine with me, but please add this part in your commit message, so it is clear that is a known change in behavior (i.e. allowing more than 1 outstanding PHY update at a time) when both the stack and the application attempts to change the PHY at the same time.

rugeGerritsen replied:
Yes, this was intentional. Let me update the commit message if @joerchan and others agree

joerchan replied:
I won't object, but maybe check with @alwa-nordic @seko-nordic as well?

Thalley replied:
@alwa-nordic seems to approve it as well :)

rugeGerritsen commented:
@jhedberg, @Vudentz  can you have a look at this PR?
